api/utils/blobaccess/chunked/access.go

```diff
@@ -0,0 +1,75 @@
+package chunked
+
+import (
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/mandelsoft/vfs/pkg/vfs"
+
+	"ocm.software/ocm/api/utils"
+	"ocm.software/ocm/api/utils/accessio"
+	"ocm.software/ocm/api/utils/blobaccess"
+	"ocm.software/ocm/api/utils/blobaccess/bpi"
+	"ocm.software/ocm/api/utils/mime"
+)
+
+func newChunk(r io.Reader, fss ...vfs.FileSystem) (bpi.BlobAccess, error) {
+	t, err := blobaccess.NewTempFile("", "chunk-*", fss...)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = io.Copy(t.Writer(), r)
+	if err != nil {
+		t.Close()
+		return nil, err
+	}
+	return t.AsBlob(mime.MIME_OCTET), nil
+}
+
+// ChunkedBlobSource provides a sequence of
+// bpi.BlobAccess objects.
+type ChunkedBlobSource interface {
```

Comment:
Couldn't this be implemented with `io.LimitedReader` instead of writing our own `ChunkedReader`?

Reply:
yes and no, the `LimitedReader` stops after N bytes, but cannot be continued.
The `ChunkedReader` offers a `Next` method to continue reading the next chunk after an EOF.
Additionally, it provides a check, whether the actual chunk is already done.

Reply:
but since we wrap the chunk readers next in the blob accesses next, couldn't we set up the consecutive limited reader in the blob accesses next?

Reply:
Unfortunately the Limited Reader does not provide information about the final EOF. If the number of bytes read, when EOF is provided, gives this hint, that the steam has been finished. But otherwise we always have to cover the case that the size is a multiple of the chunk size.  Therefore, we have to look ahead in the original stream.

This has been separated in a LookAheadReader now, which significantly reduces the complexity in the ChunkedReader.